### DISCUSSION TITLE
Git Worktree Functionality

### DISCUSSION THREAD

**Discussion by joshmedeski** [joshmedeski/sesh#69] [Ideas]
Here are some potential ideas for integrating sesh more with git worktrees

- Upon activation, retrieves and lists all worktrees that are associated with the current sessions (bare) git repository, making it easier to switch between them
- Be able to switch between “last worktree”, similar to switching between last sessions within tmux

**Comment by joshmedeski** [comment]
I personally find it really easy to jump between worktrees with sesh as-is. But if someone is looking for a more specific feature I'd like to make sure to balance simplicity and usefulness. The real power in sesh is having one global shortcut and being able to jump to any project at any time. My main concern is creating additional contextual subcommands will slow down the user, not speed them up.

**Comment by artsbentley** [comment]
One of the main issues with sesh and worktrees is that it is a lot of mental overhead to remember the existing worktrees when working on many different projects at once, and the effort it takes to switch between them. 
When I am inside a certain project I would value a way to easily switch back and forth the last active worktrees, without "losing" my last tmux session, as I would still like to switch between two projects easily,  (for this to work the sqlite implementation from other discussions might be needed)

I do agree it might be out of scope for sesh, and more worthy of it's own dedicated CLI application. However, the implementation into sesh would be relatively easy.

**Reply by joshmedeski** [reply]
For now, I've decided to avoid introducing sqlite to the project.

Have you tried [ThePrimeagen's neovim plugin](https://github.com/ThePrimeagen/git-worktree.nvim)?

Perhaps you can use sesh to open the session, then this plugin to navigate worktrees.

**Reply by artsbentley** [reply]
I've tried prime's plugin, but I do really enjoy the seperation of worktrees on a tmux session level, rather than switching within my code editor. 
I've also noticed his plugin doesnt work that great with bare repositories that have an uncommon directory layout (where the gitdir might point to .bare and worktrees live at the root of the directory)

For now I've created a simple bash script that does 90% of what i want it to and have it binded it to it's own keymap within my terminal, much like with sesh. I'll post it here so you and others could try it out if interested, maybe it would be a good way to find out if the feature could exist within sesh or not.

`note the shell is hardcoded to zsh for my use-case`

``` bash
#!/bin/bash

# Get a list of existing worktrees, filtering out those ending with ".bare"
worktrees=$(git worktree list | awk '{print $1}' | grep -v '\.bare$')

# Check if there are any worktrees
if [[ -z "$worktrees" ]]; then
	echo "No worktrees found in the repository."
	exit 1
fi

# List the existing worktrees using fzf for selection
selected_worktree=$(echo "$worktrees" | fzf +m --prompt="Select a worktree:")

# Check if the selected worktree exists
if [[ -z "$selected_worktree" ]]; then
	echo "No worktree selected. Exiting."
	exit 1
fi

# Extract the last two directories from the selected worktree path
session_name=$(basename $(dirname "$selected_worktree"))/$(basename "$selected_worktree")

# Check if a tmux session exists for the selected worktree
if tmux has-session -t "$session_name" 2>/dev/null; then
	# If session exists, switch to it
	tmux switch-client -t "$session_name"
else
	# If session doesn't exist, create a new tmux session for the selected worktree
	tmux new-session -d -s "$session_name" "cd '$selected_worktree'; exec zsh"
	tmux switch-client -t "$session_name"
fi

```

**Reply by joshmedeski** [reply]
Got it, so here's a proposal. We could add a `--dir` flag to the list command that can scan a given directory and scan it for available branches (or a certain number of children directories)

```sh
sesh list --dir ~/code/worktree-example
```

How would you want this to work? I could see being able to use the currently attached session could work too.

Let me know your thoughts.
